pyomo/contrib/cspline_external/cspline_parameters.py

```diff
@@ -313,3 +359,71 @@ def yxx_endpoint_eqn(blk, s):
         else:
             j = s
         return _fxx_cubic(m.x[j], m.alpha, s) == 0
+
+
+def add_decreasing_constraints(m, tol=0):
+    """If the objective form of the parameter calculation is used, the
+    data and the spline don't need to match exactly, and we can add
+    constraints on the derivatives that they are negative at the knots.
+
+    This doesn't necessarily mean the cubic spline function is always
+    decreasing, since the segments are cubic, but you can either check the
+    resulting curve or pair it with convex or concave constraints.
+    """
+
+    @m.Constraint(m.knt_idx)
+    def yx_ineq(blk, k):
+        if k >= len(m.knt_idx):
+            s = k - 1
+        else:
+            s = k
+        return _fx_cubic(m.x[k], m.alpha, s) <= -tol
```

Comment:
It would be good for these functions to be declared at the module scope so that they would be picklable.  Given the use of the `tol` argument, maybe they should be functors:
```python
class _yx_ineq(object):
    def __init__(self, tol):
        self.tol = tol

    def __call__(self, m, k):
        if k >= len(m.knt_idx):
            s = k - 1
        else:
            s = k
        return _fx_cubic(m.x[k], m.alpha, s) <= -self.tol

def add_decreasing_constraints(m, tol=0):
    m.yx_ineq = Constraint(m.knt_idx, rule=_yx_ineq(tol))
```